### PR TITLE
[1LP][RFR] Adding NEW TEST: VMware WebMKS Console

### PR DESCRIPTION
**Pull request by kedark3**
Purpose or Intent
=================

__Adding tests__ for feature VMware WebMKS Console which should work with vsphere6 and vsphere6.5.

{{ pytest: cfme/tests/infrastructure/test_webmks_console.py --use-provider vsphere6-nested --use-provider vsphere65-nested}}

### CODE REVIEW

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+        - The console's status will be checked.
+        - A command that creates a file will be sent through the console.
+        - Using ssh we will check that the command worked (i.e. that the file
+          was created.
+    """
+    console_vm_username = credentials[provider.data.templates.get('console_template')
+                            ['creds']].get('username')
+    console_vm_password = credentials[provider.data.templates.get('console_template')
+                            ['creds']].get('password')
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_webmks_console.py]: Is the 'console_template' key guaranteed to be always there? This will fail with TypeError if not.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    @request.addfinalizer
+    def _delete_vm():
+        try:
+            vm_obj.delete_from_provider()
+        except Exception:
+            logger.warning("Failed to delete vm `{}`.".format(vm_obj.name))
+
+    vm_obj.create_on_provider(timeout=2400, find_in_cfme=True, allow_skip="default")
+    return vm_obj
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **OPTIONAL**
You could make yield_fixture instead.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    navigate_to(appliance.server, 'Server')
+
+    settings_pg = VMwareConsoleSupport(
+        appliance=appliance,
+        console_type='VMware WebMKS',
+    )
+    settings_pg.update()
+
+
+@pytest.fixture(scope="session")
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
```python
@pytest.yield_fixture(scope="session")
```

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        with ssh.SSHClient(hostname=vm_obj.ip_address, username=console_vm_username,
+                password=console_vm_password) as ssh_client:
+            # if file was created in previous steps it will be removed here
+            # we will get instance of SSHResult
+            # Sometimes Openstack drops characters from word 'blather' hence try to remove
+            # file using partial file name. Known issue, being worked on.
+            command_result = ssh_client.run_command("rm blather", ensure_user=True)
+            assert command_result
+    finally:
+        vm_console.close_console_window()
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_webmks_console.py]: The file is more-or-less the same as cfme/tests/cloud_infra_common/test_html5_vm_console.py
What are the plans for refactoring wrt reusing shared functionality?

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    """
+    console_vm_username = credentials[provider.data.templates.get('console_template')
+                            ['creds']].get('username')
+    console_vm_password = credentials[provider.data.templates.get('console_template')
+                            ['creds']].get('password')
+
+    vm_obj.open_console(console='VM Console', invokes_alert=True)
+    assert vm_obj.vm_console, 'VMConsole object should be created'
+    vm_console = vm_obj.vm_console
+    try:
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
Why is this required? You could just do:
```python
request.addfinalizer(vm_console.close_console_window)
```

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    """
+    roles = configuration.get_server_roles()
+    if 'websocket' in roles and not roles['websocket']:
+        logger.info('Enabling the websocket role to allow console connections')
+        roles['websocket'] = True
+        configuration.set_server_roles(**roles)
+        yield
+    roles['websocket'] = False
+    logger.info('Disabling the websocket role to avoid intrusive popups')
+    configuration.set_server_roles(**roles)
```
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: Yes, websocket will be disabled if it was already enabled before, because @psav and I had a discussion when I was creating these tests and I found out that provider setup fixture was disabling the websocket role, I asked pete if I could stop that from happening, he suggested me to create this yield fixture that would enable the websocket which is disabled by the provider fixture and then disable it back when my tests are complete.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 1 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: Yes, console_template key is guaranteed to be there on all the providers which support these tests, and hence it should not fail. We have a "webmks_console" flag to filter which providers are supported in the cfme_data.yaml and those providers which are supported will always fulfill the requirement, unless somebody deletes that key from cfme_data.yaml.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 4 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: I agree its similar file. But haven't given much thought to it. Primary aim is to get the test scenario covered in this prototype. And once this is done, I can proceed to reducing the duplication. Suggestions welcome. :smile:

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 6 above)
Comment (mkoura) [inline cfme/tests/infrastructure/test_webmks_console.py]: But why not to revert it back to the state where it was before the fixture execution? Why always disabling it, even in situation when it was enabled?

Comment (mkoura) [inline cfme/tests/infrastructure/test_webmks_console.py]: @kedark3 explained to me the reasons behind this in detail offline and I agreed than in current situation (other fixture is setting it to disabled) it's safer to not touch it during teardown.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+    # Enter Username:
+    vm_console.send_keys(console_vm_username)
+
+    assert vm_console.wait_for_text(text_to_find="Password", timeout=200),\
+        "VM Console didn't prompt for Password"
+    # Enter Password:
+    vm_console.send_keys("{}\n".format(console_vm_password))
+
+    time.sleep(5)  # wait for login to complete
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: Is there way how to know  when login is completed?

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+    logger.info("Wait to get the '$' prompt")
+
+    vm_console.wait_for_text(text_to_find=provider.data.templates.get('console_template')
+                    ['prompt_text'], timeout=200)
+
+    # create file on system
+    vm_console.send_keys("touch blather\n")
+    logger.info("Wait 5 sec for file creation")
+    time.sleep(5)  # Wait for file creation
```
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**
This must be replaced by `wait_for`

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 7 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: @quarckster  There is not really a way to know this, time.sleep is needed because after I enter username and password, I need to wait at least a sec or two before I get either command prompt or "Login Incorrect" message. So I can not think of anything besides time.sleep in this situation. You might say I can wait_for _command prompt_ or _"Login Incorrect"_ message. But for Image to Text recognition finding out command prompt correctly is a challenge. And if I detect _"Login Incorrect"_ I will then need to attempt to login again. Hence I just replaced all that login with validate_login which validates if login went through and if not, attempt to re-login. Any suggestions to make it even better are welcome. Thanks :)

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 8 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: This is simple blind sleep before SSH Client can go in and try to remove the file. I want some buffer to make sure command had enough time to create a file. Would you still suggest changing this with something like 
`wait_for(vm_console.send_keys("touch blather"), num_sec=5, delay=5)` ?

Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: You should check if the file is created in the vm via ssh_client

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 7 above)
Comment (quarckster) [inline cfme/tests/infrastructure/test_webmks_console.py]: I believe it can be checked via `ssh_client` and `who` command. I would investigate such approach.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 8 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: I think removing the file (as done on  line 168 below) makes sure the file was created. That is the test here, i.e. if console is functioning correctly and capable of taking the input, file should be created(within couple of seconds) after I sent touch command, and when I run "rm blather" via SSH, if the console was functioning correctly, I will have the file present on VM and "rm blather" will work correctly. Otherwise it will fail to find the file and remove it causing failure. Making sure the file is created via ssh_client seems extra work which IMHO not required. Please provide your thoughts.

File: cfme/common/vm.py [hunk 9]
```diff
@@ -374,7 +374,10 @@ def open_console(self, console='VM Console', invokes_alert=False, cancel=False):
         view = navigate_to(self, 'Details')
 
         # Click console button given by type
-        view.toolbar.access.item_select(console)
+        if invokes_alert:
```
Comment (psav) [inline cfme/common/vm.py]: **REQUIRED**: Given this is ```False``` by default and I'm thinking ```True``` if not, why can't we just do

```view.toolbar.access.item_select(console, handle_alert=invokes+alert)```

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+@pytest.fixture(scope="module")
+def configure_vmware_console_for_test(appliance, provider):
+    """Configure VMware Console to use VNC which is what is required for the HTML5 console."""
+    navigate_to(appliance.server, 'Server')
+
+    settings_pg = VMwareConsoleSupport(
+        appliance=appliance,
+        console_type='VMware WebMKS',
+    )
+    settings_pg.update()
```
Comment (psav) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**: Should we unset this afterwards?

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+@pytest.yield_fixture(scope="session")
+def configure_websocket():
+    """Enable websocket role if it is disabled.
+
+    Currently the fixture cfme/fixtures/base.py,
+    disables the websocket role to avoid intrusive popups. It is enabled
+    here before test begins and disabled during teardown.
+    """
+    roles = configuration.get_server_roles()
```
Comment (psav) [inline cfme/tests/infrastructure/test_webmks_console.py]: **QUESTION**: Doesn't the set_roles check with the get_roles anyway? I think there is more code here than there needs to be

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 7 above)
Comment (psav) [inline cfme/tests/infrastructure/test_webmks_console.py]: **QUESTION**: I was going to ask just what you mentioned @kedark3 about checking for the two strings. Can't we just wait_for the prompt and if it doesn't show up after a certain amount of time then we know it failed?

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 8 above)
Comment (psav) [inline cfme/tests/infrastructure/test_webmks_console.py]: It is required, arbitrary sleeps are not immune to networking problems. The problem is without wait_for any sleep we put in is permanent and everytime we need to wait, we need to wait that long. Since boxes can be under load, or connections take a hwile to connect etc, I prefer the wait_for, I odn't think it's that much extra work, and meas we can retry every 10 seconds, whilst having a maximum timeout of 2minutes if we wanted. 99% of the time it'll be done within 10 seconds, but if it has to wait longer it's not big deal.

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    vm_console.wait_for_text(text_to_find=provider.data.templates.get('console_template')
+                    ['prompt_text'], timeout=200)
+
+    # create file on system
+    vm_console.send_keys("touch blather\n")
+    logger.info("Wait 5 sec for file creation")
+    time.sleep(5)  # Wait for file creation
+
+    with ssh.SSHClient(hostname=vm_obj.ip_address, username=console_vm_username,
+            password=console_vm_password) as ssh_client:
```
Comment (psav) [inline cfme/tests/infrastructure/test_webmks_console.py]: **REQUIRED**: let's call this ```vm_ssh_client``` to differentiate it from the normal ```ssh_client``` which is usually used for the appliance itself

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 7 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: Pete, unfortunately, I was not able to success tesseract to parse the image correctly and identify the prompt, but @quarckster's method seems like a good option. I think I should give it a try. Any inputs welcome.

File: cfme/tests/infrastructure/test_webmks_console.py (hunk 10 above)
Comment (kedark3) [inline cfme/tests/infrastructure/test_webmks_console.py]: Unset to what? It defaults to "VMware VMRC Plugin" when the appliance is created. Should I set it back to that? And why is it required?

**Comment by kedark3** [comment]
Currently failing with "Create_on_provider" function since there is some change in the pagination with new build 5.8.2.0. http://10.16.45.124/y7ibwMNb/artifacts/cfme/tests/infrastructure/test_webmks_console.py/test_webmks_vm_console[vsphere65-nested]/filedump-traceback.log

File: cfme/tests/infrastructure/test_webmks_console.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        return result == []
+
+    # if _validate_login() returns True, it means we did not find any of words
+    # [login, password, incorrect] on last line of console text, which implies login success
+    wait_for(func=_validate_login, timeout=300, delay=5)
+
+    logger.info("Wait to get the '$' prompt")
+
+    vm_console.wait_for_text(text_to_find=provider.data.templates.get('console_template')
+                    ['prompt_text'], timeout=200)
```
Comment (mkoura) [inline cfme/tests/infrastructure/test_webmks_console.py]: Strange indentation / line break.
